### PR TITLE
Fixed TString File

### PR DESCRIPTION
**Pull request by AdvaitDhingra**
Sorry, this is my third PR on this subject.

This patch will fix [this](https://github.com/root-project/root/issues/6748) issue.

### CODE REVIEW

**Comment by AdvaitDhingra** [comment]
@jblomer is everything in order now?

**Comment by eguiraud** [comment]
@phsft-bot build

File: core/base/src/TString.cxx [hunk 1]
```diff
@@ -238,13 +246,21 @@ TString::~TString()
...
+   }
+   if (nchar > capacity){
+      nchar = capacity;  
+   }
    if (capacity > MaxSize()) {
       Error("TString::Init", "capacity too large (%d, max = %d)", capacity, MaxSize());
       capacity = MaxSize();
-      if (nchar > capacity)
-         nchar = capacity;
-   }
```
Comment (eguiraud) [inline core/base/src/TString.cxx:246]: I think we needed this `}` :smile:

Comment (AdvaitDhingra) [reply core/base/src/TString.cxx:246]: Oh haha

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by AdvaitDhingra** [comment]
Did I add the right semicolon? I am unsure

**Comment by AdvaitDhingra** [comment]
@phsft-bot build

**Comment by eguiraud** [comment]
@phsft-bot build please

**Comment by eguiraud** [comment]
@AdvaitDhingra you can also compile ROOT locally and test your changes, instructions are at https://root.cern/install/build_from_source/

**Comment by AdvaitDhingra** [comment]
> @AdvaitDhingra you can also compile ROOT locally and test your changes, instructions are at https://root.cern/install/build_from_source/

Ah ok

File: core/base/test/TStringTest.cxx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
```
Comment (eguiraud) [inline core/base/test/TStringTest.cxx:2]: no need for this include

Comment (AdvaitDhingra) [reply core/base/test/TStringTest.cxx:2]: Both of them?

File: core/base/test/TStringTest.cxx [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+#include "TString.h"
+
+TEST(TString, Basics)
+{
+   TString n("Test", -5);
```
Comment (eguiraud) [inline core/base/test/TStringTest.cxx:8]: we should also capture the Error, otherwise it will end up in our test logs.

you need to `#include "ROOTUnitTestSupport.h"` and then you can execute code that should error out such as this and also line 13 within `ROOT_EXPECT_ERROR`, e.g.

```cpp
TString *s = nullptr;
ROOT_EXPECT_ERROR(s = new TString("Test", -5), "TString::TString", "<error message goes here>");
```

File: core/base/test/TStringTest.cxx (hunk 2 above)
Comment (eguiraud) [reply core/base/test/TStringTest.cxx:2]: no just `gmock`, `gtest.h` brings in the `TEST` and `EXPECT_*` macros that we use

**Comment by eguiraud** [comment]
@pcanal besides the two small things I commented above the changes look reasonable, maybe give it a quick look in case you don't think this is the correct way to fix #6748

**Comment by AdvaitDhingra** [comment]
Is the error handling correct?

File: core/base/test/TStringTest.cxx [hunk 4]
```diff
@@ -1,15 +1,18 @@
 #include "gtest/gtest.h"
-#include "gmock/gmock.h"
+
+#include "ROOTUnitTestSupport.h"
 
 #include "TString.h"
 
 TEST(TString, Basics)
 {
    TString n("Test", -5);
```
Comment (eguiraud) [inline core/base/test/TStringTest.cxx:9]:
```suggestion
   TString *s = nullptr;
```

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by AdvaitDhingra** [comment]
> Build failed on ROOT-fedora30/cxx14.
> Running on root-fedora30-1.cern.ch:/build/workspace/root-pullrequests-build
> [See console output](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/106458/console).
> 
> ### Errors:
> * [2021-02-11T17:52:31.264Z] FAILED: core/base/test/CMakeFiles/CoreBaseTests.dir/TStringTest.cxx.o
> * [2021-02-11T17:52:31.570Z] /build/workspace/root-pullrequests-build/root/core/base/test/TStringTest.cxx:10:21: error: ‘n’ was not declared in this scope; did you mean ‘yn’?

Fixed this.

File: core/base/test/TStringTest.cxx [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include "gtest/gtest.h"
+
+#include "ROOTUnitTestSupport.h"
+
+#include "TString.h"
+
+TEST(TString, Basics)
+{
+   TString *s = nullptr;
+   EXPECT_STREQ("", s);
```
Comment (eguiraud) [inline core/base/test/TStringTest.cxx:10]: ah, this line doesn't belong here. The tricky part is that `TString *s = nullptr` does _not_ construct a `TString` object, it constructs a _pointer_ to a `TString` object, and that pointer is "null" at this point (i.e. it points to nothing).

Later, `s = new TString` actually creates a `TString` object and makes `s` point to it. Python doesn't have pointers, C++ does!

Comment (AdvaitDhingra) [reply core/base/test/TStringTest.cxx:10]: Ohh okay.

Comment (AdvaitDhingra) [reply core/base/test/TStringTest.cxx:10]: So instead of 
```
TString *s = nullptr;
```
I write:
```
s = new TString("Test", -5);
```
?

**Comment by eguiraud** [comment]
@phsft-bot build

File: core/base/test/TStringTest.cxx [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+#include "ROOTUnitTestSupport.h"
+
+#include "TString.h"
+
+TEST(TString, Basics)
+{
+   TString *s = nullptr;
+   ROOT_EXPECT_ERROR(s = new TString("Test", -5), "TString::TString", "Negative length!");
+   EXPECT_STREQ("", s);
```
Comment (eguiraud) [inline core/base/test/TStringTest.cxx:11]: ok probably the last thing, this line should be removed and a `delete s` should be added (it destroys the TString object we created with `new`, which is leaked otherwise. pointers...):

```suggestion
   delete s;
```

Comment (AdvaitDhingra) [reply core/base/test/TStringTest.cxx]: Okay. I committed the change

**Comment by AdvaitDhingra** [comment]
What is the error? I can't open the console output

**Comment by eguiraud** [comment]
Ah, sorry @AdvaitDhingra . Here it is:

```
[ RUN      ] TString.Basics
Error in <TString::Replace>: Negative number of replacement characters!
/Users/sftnight/build/workspace/root-pullrequests-build/root/test/unit_testing_support/ROOTUnitTestSupport.h:79: Failure
Value of: Seen
Expected: starts with "Error in <TString::Append>: "
  Actual: "Error in <TString::Replace>: Negative number of replacement characters!\n"
/Users/sftnight/build/workspace/root-pullrequests-build/root/test/unit_testing_support/ROOTUnitTestSupport.h:86: Failure
Value of: Seen
Expected: is equal to "Error in <TString::Append>: Negative length!\n"
  Actual: "Error in <TString::Replace>: Negative number of replacement characters!\n"
```

(you can also run the test locally if you compile ROOT with your changes).

Anyway, looks like our expected error message does not match the actual error message.

File: core/base/src/TString.cxx [hunk 7]
```diff
@@ -362,7 +379,11 @@ TString& TString::operator=(const TSubString &substr)
 TString& TString::Append(char c, Ssiz_t rep)
 {
    if (!rep) return *this;
-
+   
+   if (rep < 0){
+      Error("TString& TString::Append", "Negative length!");
```
Comment (eguiraud) [inline core/base/src/TString.cxx:384]: just notices:

```suggestion
      Error("TString::Append", "Negative length!");
```

**Comment by eguiraud** [comment]
I am not 100% where the `TString::Replace` comes from, I'd need to checkout these changes and run the test

**Comment by AdvaitDhingra** [comment]
> Ah, sorry @AdvaitDhingra . Here it is:
> 
> ```
> [ RUN      ] TString.Basics
> Error in <TString::Replace>: Negative number of replacement characters!
> /Users/sftnight/build/workspace/root-pullrequests-build/root/test/unit_testing_support/ROOTUnitTestSupport.h:79: Failure
> Value of: Seen
> Expected: starts with "Error in <TString::Append>: "
>   Actual: "Error in <TString::Replace>: Negative number of replacement characters!\n"
> /Users/sftnight/build/workspace/root-pullrequests-build/root/test/unit_testing_support/ROOTUnitTestSupport.h:86: Failure
> Value of: Seen
> Expected: is equal to "Error in <TString::Append>: Negative length!\n"
>   Actual: "Error in <TString::Replace>: Negative number of replacement characters!\n"
> ```
> 
> (you can also run the test locally if you compile ROOT with your changes).
I'm having some trouble with that. What id the <sourcedir>? When I point it towards the CMakeLists.txt file in the test folder, it thows me an error. Is this the right folder?
> 
> Anyway, looks like our expected error message does not match the actual error message.

**Comment by AdvaitDhingra** [comment]
> I am not 100% where the `TString::Replace` comes from, I'd need to checkout these changes and run the test

Yeah, I'm wondering that too

**Comment by eguiraud** [comment]
You really need to compile all of ROOT, I'm afraid (well, at least the core parts, but at that point you might as well compile all of it). So from the build directory you create, you need to point cmake to ROOT's top-level source directory.

In case you don't manage, I'll take a look when I can

**Comment by AdvaitDhingra** [comment]
> You really need to compile all of ROOT, I'm afraid (well, at least the core parts, but at that point you might as well compile all of it). So from the build directory you create, you need to point cmake to ROOT's top-level source directory.
> 
> In case you don't manage, I'll take a look when I can

Okay, I'll give it a try

**Comment by AdvaitDhingra** [comment]
I tried a.Append("s", -5) in my console, and it throws me the "TString::Replace>: Negative number of replacement characters" error. Should I change the error message in my test file?

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by AdvaitDhingra** [comment]
Seems to work on Linux systems and OSX

**Comment by eguiraud** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by AdvaitDhingra** [comment]
@eguiraud can you link this PR to the issue, i.e. when this PR is merged the Issue is resolved?

**Comment by AdvaitDhingra** [comment]
What is the error?

**Comment by eguiraud** [comment]
@AdvaitDhingra almost done, please run `git clang-format master` and commit the formatting fixes, there are some whitespace issues and similar and `git clang-format master` fixes them for you, taking care to only touch what your branch changed w.r.t. the `master` branch.

@pcanal can you give us your ok please?

@bellenot I don't understand the windows build error, can you confirm that it's unrelated to this patch please?

**Comment by AdvaitDhingra** [comment]
> @AdvaitDhingra almost done, please run `git clang-format master` and commit the formatting fixes, there are some whitespace issues and similar and `git clang-format master` fixes them for you, taking care to only touch what your branch changed w.r.t. the `master` branch.
How can I format the commits that I have already pushed?

**Comment by eguiraud** [comment]
> How can I format the commits that I have already pushed?

no need, just add a commit that fixes the formatting of your branch (e.g. with the command above). We will squash the 12 commits in this PR together into a single commit anyway when merging into ROOT.

>  it tells me that clang-format is not a git command.

uhm, you might need to install _something_, depending on your system :sweat_smile:

**Comment by AdvaitDhingra** [comment]
> > How can I format the commits that I have already pushed?
> 
> no need, just add a commit that fixes the formatting of your branch (e.g. with the command above). We will squash the 12 commits in this PR together into a single commit anyway when merging into ROOT.
> 
> > it tells me that clang-format is not a git command.
> 
> uhm, you might need to install _something_, depending on your system sweat_smile

Ah ok. Thanks

**Comment by bellenot** [comment]
> @bellenot I don't understand the windows build error, can you confirm that it's unrelated to this patch please?

I don't understand why it's happening here and not in other builds, but I guess it's unrelated (at least I hope)

**Comment by AdvaitDhingra** [comment]
What does the build error message say? I can't read it

**Comment by bellenot** [comment]
> What does the build error message say? I can't read it

```
12:53:51  MSVCRT.lib(tncleanup.obj) : error LNK2005: "struct __type_info_node __type_info_root_node" (?__type_info_root_node@@3U__type_info_node@@A) already defined in libCling.lib(libCling.dll) [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
12:53:51       Creating library C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.lib and object C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.exp
12:53:51  C:\build\workspace\root-pullrequests-build\build\core\base\test\Release\CoreBaseTests.exe : fatal error LNK1169: one or more multiply defined symbols found [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
```

**Comment by AdvaitDhingra** [comment]
> > What does the build error message say? I can't read it
> 
> ```
> 12:53:51  MSVCRT.lib(tncleanup.obj) : error LNK2005: "struct __type_info_node __type_info_root_node" (?__type_info_root_node@@3U__type_info_node@@A) already defined in libCling.lib(libCling.dll) [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
> 12:53:51       Creating library C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.lib and object C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.exp
> 12:53:51  C:\build\workspace\root-pullrequests-build\build\core\base\test\Release\CoreBaseTests.exe : fatal error LNK1169: one or more multiply defined symbols found [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
> ```

Thanks. I have no clue what it means

**Comment by bellenot** [comment]
> 
> Thanks. I have no clue what it means

Don't worry, I know what that means

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by eguiraud** [comment]
Looks good to me. Let's build one last time, then we wait for @pcanal to give his final ok.

File: core/base/src/TString.cxx [hunk 8]
```diff
@@ -238,11 +246,20 @@ TString::~TString()
...
+      nchar = 0;
+   }
+   if (nchar > capacity) {
+      nchar = capacity;
+   }
    if (capacity > MaxSize()) {
       Error("TString::Init", "capacity too large (%d, max = %d)", capacity, MaxSize());
       capacity = MaxSize();
-      if (nchar > capacity)
-         nchar = capacity;
```
Comment (eguiraud) [inline core/base/src/TString.cxx:245]: why did you move this code out of the outer `if`?

Comment (AdvaitDhingra) [reply core/base/src/TString.cxx:245]: I'm not sure 😅 . Should I put it back into the capacity > MaxSize() if-statement?

Comment (eguiraud) [reply core/base/src/TString.cxx:245]: yes please :smile:

File: core/base/src/TString.cxx [hunk 9]
```diff
@@ -254,12 +254,12 @@ char *TString::Init(Ssiz_t capacity, Ssiz_t nchar)
       Error("*TString::Init", "Negative length!");
       nchar = 0;
    }
-   if (nchar > capacity) {
-      nchar = capacity;
-   }
    if (capacity > MaxSize()) {
       Error("TString::Init", "capacity too large (%d, max = %d)", capacity, MaxSize());
       capacity = MaxSize();
+      if (nchar > capacity) {
```
Comment (AdvaitDhingra) [inline core/base/src/TString.cxx:260]: Just realized that the curly brackets are unnecessary. Should I remove them or just keep it like this?

Comment (eguiraud) [reply core/base/src/TString.cxx:260]: Keep it like it is please (less noise in the diffs).

Comment (AdvaitDhingra) [reply core/base/src/TString.cxx:260]: Okay

Comment (eguiraud) [reply core/base/src/TString.cxx:260]: argh sorry, I was mistaken, there were no brackets originally! So indeed let's keep it without -- sorry for the back and forth!

Comment (AdvaitDhingra) [reply core/base/src/TString.cxx:260]: No problem. I'll change it

File: core/base/src/TString.cxx [hunk 10]
```diff
@@ -363,6 +379,10 @@ TString& TString::Append(char c, Ssiz_t rep)
 {
    if (!rep) return *this;
 
+   if (rep < 0) {
+      Error("TString::Append", "Negative length!");
+      Zero();
+   }
```
Comment (pcanal) [inline core/base/src/TString.cxx:385]:
```suggestion
   if (rep < 0) {
      Error("TString::Append", "Negative length!");
      return *this
   }
```
Since there length is negative, there is technically nothing to add and doing an "Append" should not result (even in fail mode) with the string being emptied.

**Comment by AdvaitDhingra** [comment]
For some reason the build keeps failing on windows and travis-ci. Not sure about windows, but the travis-ci has continuously failed since the third PR. Any ideas?

File: core/base/src/TString.cxx [hunk 11]
```diff
@@ -363,6 +379,10 @@ TString& TString::Append(char c, Ssiz_t rep)
 {
    if (!rep) return *this;
 
+   if (rep < 0) {
+      Error("TString::Append", "Negative length!");
+      return *this
```
Comment (eguiraud) [inline core/base/src/TString.cxx:384]:
```suggestion
      return *this;
```

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by AdvaitDhingra** [comment]
Thank you all for your patience. A PR probably doesn't usually take this long 😅

**Comment by eguiraud** [comment]
Travis complains about a potential memory leak but I think it's a false positive (@pcanal can confirm/deny). For the Windows failure we need @bellenot 's green light to merge, or his advice on how to fix it.

**Comment by bellenot** [comment]
It looks like the error on Windows is due to this PR, but I don't understand why (not yet). I'll have to test this PR myself

**Comment by AdvaitDhingra** [comment]
What is the error?

**Comment by pcanal** [comment]
@phsft-bot build just on ROOT-debian10-i386/cxx14

**Comment by AdvaitDhingra** [comment]
> On windows:
> 
> ```
> 
> 11:06:03  MSVCRT.lib(tncleanup.obj) : error LNK2005: "struct __type_info_node __type_info_root_node" (?__type_info_root_node@@3U__type_info_node@@A) already defined in libCling.lib(libCling.dll) [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
> 
> 11:06:03    ParallelTest.cxx
> 
> 11:06:03       Creating library C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.lib and object C:/build/workspace/root-pullrequests-build/build/core/base/test/Release/CoreBaseTests.exp
> 
> 11:06:03    PaulTest.cxx
> 
> 11:06:03    REveTreeTools.cxx
> 
> 11:06:03  C:\build\workspace\root-pullrequests-build\build\core\base\test\Release\CoreBaseTests.exe : fatal error LNK1169: one or more multiply defined symbols found [C:\build\workspace\root-pullrequests-build\build\core\base\test\CoreBaseTests.vcxproj]
> 
> ```

I mean on debian

**Comment by AdvaitDhingra** [comment]
I just built this branch on my machine (Ubuntu 20.04) and everything works fine. Strange, because Ubuntu is Debian based.

**Comment by pcanal** [comment]
It is a strange error looking like a transient infrastructure issue, consequently I am running that build.

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by AdvaitDhingra** [comment]
What's happening

**Comment by pcanal** [comment]
I have run *all* the test and *all* the platforms and some are failing for reasons unrelated to this PR.

**Review by pcanal — APPROVED** [review]

**Comment by eguiraud** [comment]
Congrats @AdvaitDhingra !

**Comment by AdvaitDhingra** [comment]
> Congrats @AdvaitDhingra !

Thank you 😊. And thank you to everyone's help (especially @eguiraud). I have learned a lot and will continue to contribute to this project 😀. Have a nice weekend.

**Comment by bellenot** [comment]
@pcanal FYI, this broke ROOT master on Windows...

**Comment by AdvaitDhingra** [comment]
> @pcanal FYI, this broke ROOT master on Windows...

Really? Because @pcanal said that it was unrelated to this PR. Do you know what failed? Maybe someone can open an issue.

**Comment by bellenot** [comment]
> > @pcanal FYI, this broke ROOT master on Windows...
> 
> Really? Because @pcanal said that it was unrelated to this PR. Do you know what failed? Maybe someone can open an issue.

Yes, really. But I'll fix it

**Comment by eguiraud** [comment]
Don't worry @AdvaitDhingra , not your fault and work in progress :)

**Comment by AdvaitDhingra** [comment]
> Don't worry @AdvaitDhingra , not your fault and work in progress :)

Okay, thanks for the reassurance 😄

**Comment by pcanal** [comment]
@bellenot Sorry, I indeed misread the windows error :(

**Comment by bellenot** [comment]
> @bellenot Sorry, I indeed misread the windows error :(

@pcanal no problem, it should be fixed now.
